pkg/objstore/azure/azure.go

```diff
@@ -0,0 +1,300 @@
+	logger       log.Logger
+	containerURL blob.ContainerURL
+	config       *Config
+	opsTotal     *prometheus.CounterVec
+	closer       blobReadCloser
+}
+
+// Validate checks to see if any of the s3 config options are set.
+func (conf *Config) Validate() error {
+	if conf.StorageAccountName == "" ||
+		conf.StorageAccountKey == "" {
+		return errors.New("invalid Azure storage configuration")
+	}
+	return nil
+}
+
+// NewBucket returns a new Bucket using the provided Azure config.
+func NewBucket(logger log.Logger, azureConfig []byte, reg prometheus.Registerer, component string) (*Bucket, error) {
+
+	level.Debug(logger).Log("msg", "Creating new Azure bucket connection", "component", component)
+
+	var conf Config
+	if err := yaml.Unmarshal(azureConfig, &conf); err != nil {
+		return nil, err
+	}
+
+	err := conf.Validate()
+	if err != nil {
+		return nil, err
+	}
+
+	ctx := context.Background()
+
+	container, err := createContainer(ctx, conf.StorageAccountName, conf.StorageAccountKey, conf.ContainerName)
+	if err != nil {
+		if err.(blob.StorageError).ServiceCode() == "ContainerAlreadyExists" {
+			level.Debug(logger).Log("msg", "Using existing Azure blob container", "address", container)
+			level.Debug(logger).Log("msg", "Getting connection to existing Azure blob container", "container", conf.ContainerName)
+			container, err = getContainer(ctx, conf.StorageAccountName, conf.StorageAccountKey, conf.ContainerName)
+			if err != nil {
+				level.Error(logger).Log("msg", "Cannot get existing Azure blob container", "address", container)
+				return nil, err
+			}
+		} else {
+			level.Error(logger).Log("msg", "Error creating Azure blob container", "address", container)
+			return nil, err
+		}
+	} else {
+		level.Info(logger).Log("msg", "Azure blob container successfully created", "address", container)
+	}
+
+	bkt := &Bucket{
+		logger:       logger,
+		containerURL: container,
+		config:       &conf,
+		opsTotal: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name:        "thanos_objstore_azure_storage_operations_total",
+			Help:        "Total number of operations that were executed against an Azure storage account.",
+			ConstLabels: prometheus.Labels{"storage_account": conf.StorageAccountName},
+		}, []string{"operation"}),
+	}
+
+	if reg != nil {
+		reg.MustRegister(bkt.opsTotal)
+	}
+	return bkt, nil
+}
+
+// Iter calls f for each entry in the given directory. The argument to f is the full
+// object name including the prefix of the inspected directory.
+func (b *Bucket) Iter(ctx context.Context, dir string, f func(string) error) error {
+	level.Debug(b.logger).Log("msg", "Iterating through the dir", "dir", dir)
+	b.opsTotal.WithLabelValues(opObjectsList).Inc()
+	var prefix string
+	if dir == "" {
+		prefix = ""
+	} else if !strings.HasSuffix(dir, DirDelim) {
+		prefix = dir + DirDelim
+	} else {
+		prefix = dir
+	}
+
+	list, err := b.containerURL.ListBlobsHierarchySegment(ctx, blob.Marker{}, DirDelim, blob.ListBlobsSegmentOptions{
```

Comment:
I guess there is no native iter method? This way we seem to hide non streamed logic under stream like `Iter` method. Would be nice to avoid this if possible

Reply:
There is no native Iter method, unfortunately :(